### PR TITLE
Make it possible to stream the terms matching an Automaton

### LINKED ISSUE #273: Add a `.search(automaton: fst::Automaton)` to the `TermDictionary`
This is required for `#219` (sublime text like search) and `#272` (FuzzyTermQuery).

Note that while similar, they are quite different in that 
- in `#219`, the user probably want the list of terms that matched, possibly sorted by a function of the doc frequencies and some function of the query and the term`. (See the ticket comments for some links)
- in `#272`, this is effectively a query. The user wants the list of ranked documents.

The method already exists is already implemented in the `fst` crate.
Unfortunately it does not give access to the automaton state in the resulting `Streamer`.

I [opened a ticket](https://github.com/BurntSushi/fst/issues/60) in the `fst` crate to discuss whether this feature should be added upstream.
Otherwise, the `raw` module of the `fst` crate exposes enough public API to reimplement the intersection in tantivy if required.

### PR DESCRIPTION
**Pull request by drusellers**
Closes #273

Ok, so I've started down the path, but I'm for sure lost.

1. Not sure how I would test with Automaton - still reading up on the fst code there
2. Not sure how to use the fst StreamBuilder, but it feels close on the code implementation, but I have compiler issues, which could just be about spreading the generic type love around more.

### CODE REVIEW

File: src/termdict/termdict.rs [hunk 1]
```diff
@@ -101,7 +102,7 @@ fn open_fst_index(source: ReadOnlySource) -> fst::Map {
 /// The term dictionary contains all of the terms in
 /// `tantivy index` in a sorted manner.
 ///
-/// The `Fst` crate is used to assoicated terms to their
+/// The `Fst` crate is used to assoicate terms to their
```
Comment (fulmicoton) [inline src/termdict/termdict.rs]: hehe there is still the typo :)

File: src/termdict/termdict.rs [hunk 2]
```diff
@@ -199,4 +200,13 @@ impl TermDictionary {
     pub fn stream<'a>(&'a self) -> TermStreamer<'a> {
         self.range().into_stream()
     }
+
+    pub fn search<'a, A: Automaton>(&'a self, automaton: A) -> TermStreamerBuilder<'a, A> {
+        use fst::map;
+
+        // given an Automaton and a fst::Map (fst_index)
+        // how can I generate a streambuilder?
```
Comment (fulmicoton) [inline src/termdict/termdict.rs]: I don't understand the question here???

**Comment by fulmicoton** [comment]
Can you push those to the tantivy repo as  feature branches in the future? This will make it possible for me to commit stuff.

File: src/termdict/termdict.rs (hunk 2 above)
Comment (fulmicoton) [reply src/termdict/termdict.rs]: oh gotcha:
```rust
    pub fn search<'a, A: Automaton>(&'a self, automaton: A) -> TermStreamerBuilder<'a, A> {
        let sb = self.fst_index.search(automaton);
        TermStreamerBuilder::<A>::new(self, sb)
    }
```

File: src/termdict/termdict.rs (hunk 1 above)
Comment (drusellers) [reply src/termdict/termdict.rs]: ha! that's what I get. ;p

File: src/termdict/termdict.rs (hunk 2 above)
Comment (drusellers) [reply src/termdict/termdict.rs]: Thank you.

**Comment by drusellers** [comment]
> Can you push those to the tantivy repo as feature branches in the future? This will make it possible for me to commit stuff.

Absolutely. :) Do you have any preferences for branch names in the tantivy repo?

Also, for the life of me I can't get the test to compile. :/

```
error[E0277]: the trait bound `fst_levenshtein::Levenshtein: fst::Automaton` is not satisfied
   --> src/termdict/mod.rs:459:31
    |
459 |         let range = term_dict.search(automaton).into_stream();
    |                               ^^^^^^ the trait `fst::Automaton` is not implemented for `fst_levenshtein::Levenshtein`
```

But it appears to be implemented right here.

https://github.com/BurntSushi/fst/blob/master/fst-levenshtein/src/lib.rs#L142

**Comment by drusellers** [comment]
Should I be using `fst_levenshtein::Levenshtein` to test this code or something else?

**Comment by fulmicoton** [comment]
Alternatively you can use `https://github.com/tantivy-search/levenshtein-automata` or write your own. You still have a compilation error?

**Comment by drusellers** [comment]
Yeah, same error. Then I saw your impl of Levenshtein and wondered if that is a better "Test"

**Comment by drusellers** [comment]
```
error[E0277]: the trait bound `levenshtein_automata::DFA: fst::Automaton` is not satisfied
   --> src/termdict/mod.rs:464:31
    |
464 |         let range = term_dict.search(dfa).into_stream();
    |                               ^^^^^^ the trait `fst::Automaton` is not implemented for `levenshtein_automata::DFA`
```

:(

**Comment by fulmicoton** [comment]
Hmmm.... The only think I can think of, is version of fst conflicting for some reason.
Let me try on my computer.

**Comment by fulmicoton** [comment]
Yeah it compiles on my computer. Maybe try `rm -f Cargo.lock`?

**Comment by drusellers** [comment]
Ok, did a cargo clean, cargo update - no love. I could try updating `fst` to `0.3.0`

**Comment by drusellers** [comment]
@fulmicoton sweet. one step in the right direction. :)

**Comment by drusellers** [comment]
`cargo 1.26.0 (0e7c5a931 2018-04-06)` FWIW

**Comment by drusellers** [comment]
`cargo build` is 👍 
`cargo test` is 👎 

just to be clear

**Comment by fulmicoton** [comment]
Ok `cargo test` fails on my computer as well. Let me check if I can solve that riddle.

**Comment by fulmicoton** [comment]
It works if you change tantivy's `fst` version to `0.3`.

If you want to use `tantivy`'s `levenshtein_automata` crate, be careful. 
The implementation of `fst::Automaton` is guarded by a feature called `fst_automaton`.

So 

`levenshtein_automata = {version="0.1", features=["fst_automaton"]}`

is the right way to add the dependency in tantivy's `Cargo.toml`

File: Cargo.toml [hunk 3]
```diff
@@ -28,6 +29,7 @@ serde_derive = "1.0"
 serde_json = "1.0"
 num_cpus = "1.2"
 itertools = "0.5.9"
+levenshtein_automata = {version="0.1", features=["fst_automaton"]}
```
Comment (drusellers) [inline Cargo.toml:31]: I should probably make this optional?

File: Cargo.toml [hunk 4]
```diff
@@ -17,7 +17,8 @@ byteorder = "1.0"
 lazy_static = "0.2.1"
 tinysegmenter = "0.1.0"
 regex = "0.2"
-fst = {version="0.2", default-features=false}
+fst = {version="0.3", default-features=false}
+fst-levenshtein = "0.2"
```
Comment (drusellers) [inline Cargo.toml]: Should this be optional - maybe behind a `fuzzy` feature gate until the whole feature is done?

File: src/directory/mmap_directory.rs [hunk 5]
```diff
@@ -40,9 +40,11 @@ fn open_mmap(full_path: &Path) -> result::Result<Option<MmapReadOnly>, OpenReadE
...
         return Ok(None);
     }
-    MmapReadOnly::open(&file)
-        .map(Some)
-        .map_err(|e| From::from(IOError::with_path(full_path.to_owned(), e)))
+    unsafe {
+        MmapReadOnly::open(&file)
+            .map(Some)
+            .map_err(|e| From::from(IOError::with_path(full_path.to_owned(), e)))
+    }
```
Comment (drusellers) [inline src/directory/mmap_directory.rs:47]: I don't understand why this change was required when I updated fst -> 0.3 (could be new rust stable too) - but would want someone else's eyes on it.

File: src/directory/read_only_source.rs [hunk 6]
```diff
@@ -42,7 +42,7 @@ impl ReadOnlySource {
     pub fn as_slice(&self) -> &[u8] {
         match *self {
             #[cfg(feature = "mmap")]
-            ReadOnlySource::Mmap(ref mmap_read_only) => unsafe { mmap_read_only.as_slice() },
+            ReadOnlySource::Mmap(ref mmap_read_only) => mmap_read_only.as_slice(),
```
Comment (drusellers) [inline src/directory/read_only_source.rs:45]: I don't understand why this change was required when I updated fst -> 0.3 (could be new rust stable too) - but would want someone else's eyes on it.

File: src/directory/mmap_directory.rs (hunk 5 above)
Comment (fulmicoton) [reply src/directory/mmap_directory.rs:47]: You mean the `unsafe`  ? There is a comment about it in the fst crate.
It was modified in a non-backward compatible way in `0.3`. As a general rule any change from `0.x` to `0.y` may change backward compatibility.
For version > 1.x, backward compatilibility is respected as long as the major version does not change.

File: src/termdict/streamer.rs [hunk 7]
```diff
@@ -1,18 +1,25 @@
...
+use fst::Automaton;
 use fst::{IntoStreamer, Streamer};
 use postings::TermInfo;
 use termdict::TermOrdinal;
 
-/// `TermStreamerBuilder` is an helper object used to define
+/// `TermStreamerBuilder` is a helper object used to define
 /// a range of terms that should be streamed.
-pub struct TermStreamerBuilder<'a> {
+pub struct TermStreamerBuilder<'a, A>
```
Comment (fulmicoton) [inline src/termdict/streamer.rs]: make `A=AlwaysMatch` a default

File: src/termdict/streamer.rs [hunk 8]
```diff
@@ -58,15 +65,21 @@ impl<'a> TermStreamerBuilder<'a> {
 
 /// `TermStreamer` acts as a cursor over a range of terms of a segment.
 /// Terms are guaranteed to be sorted.
-pub struct TermStreamer<'a> {
+pub struct TermStreamer<'a, A>
```
Comment (fulmicoton) [inline src/termdict/streamer.rs]: make `A=AlwaysMatch` a default

File: src/termdict/merger.rs [hunk 9]
```diff
@@ -1,11 +1,12 @@
+use fst::automaton::AlwaysMatch;
 use schema::Term;
 use std::cmp::Ordering;
 use std::collections::BinaryHeap;
 use termdict::TermOrdinal;
 use termdict::TermStreamer;
 
 pub struct HeapItem<'a> {
-    pub streamer: TermStreamer<'a>,
+    pub streamer: TermStreamer<'a, AlwaysMatch>,
```
Comment (fulmicoton) [inline src/termdict/merger.rs]: useless once you set the default `make `A=AlwaysMatch`

File: src/termdict/merger.rs [hunk 10]
```diff
@@ -44,7 +45,7 @@ pub struct TermMerger<'a> {
 impl<'a> TermMerger<'a> {
     /// Stream of merged term dictionary
     ///
-    pub fn new(streams: Vec<TermStreamer<'a>>) -> TermMerger<'a> {
+    pub fn new(streams: Vec<TermStreamer<'a, AlwaysMatch>>) -> TermMerger<'a> {
```
Comment (fulmicoton) [inline src/termdict/merger.rs]: useless once you set the default `make `A=AlwaysMatch`

File: src/termdict/mod.rs [hunk 11]
```diff
@@ -381,7 +383,7 @@ mod tests {
         let source = ReadOnlySource::from(buffer);
         let term_dictionary: TermDictionary = TermDictionary::from_source(source);
 
-        let value_list = |mut streamer: TermStreamer| {
+        let value_list = |mut streamer: TermStreamer<AlwaysMatch>| {
```
Comment (fulmicoton) [inline src/termdict/mod.rs]: useless once you set the default `make `A=AlwaysMatch`

File: src/termdict/mod.rs [hunk 12]
```diff
@@ -366,6 +366,8 @@ mod tests {
 
     #[test]
     fn test_stream_range_boundaries() {
+        use fst::automaton::AlwaysMatch;
```
Comment (fulmicoton) [inline src/termdict/mod.rs]: useless once you set the default `make `A=AlwaysMatch`

File: src/termdict/termdict.rs [hunk 13]
```diff
@@ -191,12 +193,19 @@ impl TermDictionary {
 
     /// Returns a range builder, to stream all of the terms
     /// within an interval.
-    pub fn range<'a>(&'a self) -> TermStreamerBuilder<'a> {
+    pub fn range<'a>(&'a self) -> TermStreamerBuilder<'a, AlwaysMatch> {
```
Comment (fulmicoton) [inline src/termdict/termdict.rs]: useless once you set the default `make `A=AlwaysMatch`

File: src/termdict/termdict.rs [hunk 14]
```diff
@@ -191,12 +193,19 @@ impl TermDictionary {
...
         TermStreamerBuilder::new(self, self.fst_index.range())
     }
 
     /// A stream of all the sorted terms. [See also `.stream_field()`](#method.stream_field)
-    pub fn stream<'a>(&'a self) -> TermStreamer<'a> {
+    pub fn stream<'a>(&'a self) -> TermStreamer<'a, AlwaysMatch> {
```
Comment (fulmicoton) [inline src/termdict/termdict.rs]: useless once you set the default `make `A=AlwaysMatch`

File: src/query/range_query.rs [hunk 15]
```diff
@@ -215,7 +216,7 @@ pub struct RangeWeight {
 }
 
 impl RangeWeight {
-    fn term_range<'a>(&self, term_dict: &'a TermDictionary) -> TermStreamer<'a> {
+    fn term_range<'a>(&self, term_dict: &'a TermDictionary) -> TermStreamer<'a, AlwaysMatch> {
```
Comment (fulmicoton) [inline src/query/range_query.rs]: useless once you set the default `make `A=AlwaysMatch`

File: src/directory/mmap_directory.rs (hunk 5 above)
Comment (drusellers) [reply src/directory/mmap_directory.rs:47]: Cool, since it was an unsafe change I just wanted someone else to look at it. :)

Comment (fulmicoton) [reply src/directory/mmap_directory.rs:47]: (To be really "philosophically" correct here, we should probably make the caller `unsafe`... I think this is borderline enough to not care.)

File: Cargo.toml (hunk 3 above)
Comment (fulmicoton) [reply Cargo.toml:31]: No that's ok.

File: src/termdict/termdict.rs (hunk 1 above)
Comment (fulmicoton) [reply src/termdict/termdict.rs]: can you fix the typo:

`assoicate` => `associate`

**Comment by drusellers** [comment]
@fulmicoton thank you for helping me with this. It was cool to see this work. I'm now excited to try and close out some more of the fuzzy related issues. 👍

File: src/termdict/termdict.rs [hunk 16]
```diff
@@ -199,4 +200,11 @@ impl TermDictionary {
     pub fn stream<'a>(&'a self) -> TermStreamer<'a> {
         self.range().into_stream()
     }
+
+    /// Returns a search builder, to stream all of the terms
+    /// within the Automaton
+    pub fn search<'a, A: Automaton>(&'a self, automaton: A) -> TermStreamerBuilder<'a, A> {
+        let sb = self.fst_index.search(automaton);
```
Comment (fulmicoton) [inline src/termdict/termdict.rs]: Let's avoid riddly short names like `sb`. especially where the type is non-trviail to the reader. 
`sb` => `stream_builder`.

**Comment by fulmicoton** [comment]
kudos! We're almost there.

**Comment by fulmicoton** [comment]
@drusellers  That was the hardest part, right there! I'll fill another ticket to get the next step toward a fuzzy search.

File: src/lib.rs [hunk 17]
```diff
@@ -136,9 +136,11 @@ extern crate combine;
 extern crate crossbeam;
 extern crate fnv;
 extern crate fst;
+extern crate fst_levenshtein;
```
Comment (fulmicoton) [inline src/lib.rs]: we can remove `fst_levenshtein` now?

File: Cargo.toml (hunk 4 above)
Comment (fulmicoton) [reply Cargo.toml]: we can remove `fst-levenshtein` and keep only `levenshtein-automata`. The latter does the same thing, but faster, better and stronger.

**Comment by fulmicoton** [comment]
We're almost good to go. Can you just remove the deps to `fst-levenshtein`?
